src/dbnode/storage/shard.go

```diff
@@ -1787,14 +1804,84 @@ func (s *dbShard) Bootstrap(
 	// (will be passed to series.Bootstrap() as BlockState).
 	s.bootstrapFlushStates()
 
+	multiErr := xerrors.NewMultiError()
+	bootstrapResult, err := s.loadSeries(bootstrappedSeries, true)
+	if err != nil {
+		multiErr = multiErr.Add(err)
+	}
+	s.emitBootstrapResult(bootstrapResult)
+
+	// From this point onwards, all newly created series that aren't in
+	// the existing map should be considered bootstrapped because they
+	// have no data within the retention period.
+	s.Lock()
+	s.newSeriesBootstrapped = true
+	s.Unlock()
+
+	// Find the series with no data within the retention period but has
+	// buffered data points since server start. Any new series added
+	// after this will be marked as bootstrapped.
+	s.forEachShardEntry(func(entry *lookup.Entry) bool {
+		seriesEntry := entry.Series
+		if seriesEntry.IsBootstrapped() {
+			return true
+		}
+		_, err := seriesEntry.Bootstrap(nil, series.BootstrappedBlockStateSnapshot{})
+		multiErr = multiErr.Add(err)
+		return true
+	})
+
+	// Now that this shard has finished bootstrapping, attempt to cache all of its seekers. Cannot call
+	// this earlier as block lease verification will fail due to the shards not being bootstrapped
+	// (and as a result no leases can be verified since the flush state is not yet known).
+	if err := s.cacheShardIndices(); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	s.Lock()
+	s.bootstrapState = Bootstrapped
+	s.Unlock()
+
+	return multiErr.FinalError()
+}
+
+func (s *dbShard) Load(
+	series *result.Map,
```

Comment:
Hm, wouldn't it make sense to do this in chunks instead of all at once (so we can move to streaming bootstrap?)

Reply:
Yeah I may change the interface here slightly once I start hooking it up to the repair API, but I think this is fine for now. The map can be reset / re-used.

Reply:
Sure thing, can revisit later.